File: cpp/79. Word Search.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;
typedef vector<vector<char>> vvc;
typedef vector<bool> vb;
typedef vector<vector<bool>> vvb;

class Solution {
public:
    bool dfs(int i, int x, int y, vvb &done, vvc &board, int m, int n, string &word) {
        if(word[i] != board[x][y]) return false;
        if(i==word.size()-1) return true;
        done[x][y] = true;
        int dirX[4] = {0, 0, 1, -1};
        int dirY[4] = {1, -1, 0, 0}; 
        for(int j=0; j<4; j++) {
            int nextX = x + dirX[j];
            int nextY = y + dirY[j];
            if(nextX >= 0 && nextX < m && nextY >= 0 && nextY < n && !done[nextX][nextY]) {
                if(dfs(i+1, nextX, nextY, done, board, m, n, word)) return true;
            }
        }
        done[x][y] = false;
        return false;
    }
    bool exist(vector<vector<char>>& board, string word) {
        int m = board.size(), n = board[0].size();
        vvb done = vvb(m, vb(n, false));
        for(int x=0; x<m; x++) {
            for(int y=0; y<n; y++) {
                if(dfs(0, x, y, done, board, m, n, word)) return true;
            }
        }
        return false;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; vvc board; string word; bool ans;
    
    solution = Solution();
    board = {{'A','B','C','E'},{'S','F','C','S'},{'A','D','E','E'}};
    word = "ABCCED";
    ans = solution.exist(board, word);
    assert(ans==true);

    solution = Solution();
    board = {{'A','B','C','E'},{'S','F','C','S'},{'A','D','E','E'}};
    word = "SEE";
    ans = solution.exist(board, word);
    assert(ans==true);

    solution = Solution();
    board = {{'A','B','C','E'},{'S','F','C','S'},{'A','D','E','E'}};
    word = "ABCB";
    ans = solution.exist(board, word);
    assert(ans==false);

    solution = Solution();
    board = {{'a','a'}};
    word = "aaa";
    ans = solution.exist(board, word);
    assert(ans==false);
}

File: python/79-Word Search.py
from typing import List
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        "Backtracking"
        m = len(board)
        n = len(board[0])
        visited = set()
        def get_next_poses(x, y):
            dirs = [(0, 1), (0, -1), (1, 0), (-1, 0)]
            for dir in dirs:
                next_x, next_y = x + dir[0], y + dir[1]
                if 0<=next_x<m and 0<=next_y<n and (next_x, next_y) not in visited:
                    yield next_x, next_y
        def dfs(i, x, y):
            if board[x][y] != word[i]: return False
            if i == len(word)-1: return True
            visited.add((x, y))
            for next_x, next_y in get_next_poses(x, y):
                if dfs(i+1, next_x, next_y): return True
            visited.remove((x, y))
            return False
        for x in range(m):
            for y in range(n):
                if dfs(0, x, y): return True
        return False

board = [["A","B","C","E"],
         ["S","F","C","S"],
         ["A","D","E","E"]]
word = "ABCCED"
result = Solution().exist(board, word)
assert result == True

board = [["A","B","C","E"],
         ["S","F","C","S"],
         ["A","D","E","E"]]
word = "SEE"
result = Solution().exist(board, word)
assert result == True

board = [["A","B","C","E"],
         ["S","F","C","S"],
         ["A","D","E","E"]]
word = "ABCB"
result = Solution().exist(board, word)
assert result == False

board = [["a","b"],
         ["c","d"]]
word = "acdb"
result = Solution().exist(board, word)
assert result == True